lib/android/src/main/java/com/airbnb/android/react/maps/AirMapMbTile.java

```diff
@@ -0,0 +1,139 @@
+package com.airbnb.android.react.maps;
+
+import android.content.Context;
+
+import com.google.android.gms.maps.GoogleMap;
+import com.google.android.gms.maps.model.Tile;
+import com.google.android.gms.maps.model.TileOverlay;
+import com.google.android.gms.maps.model.TileOverlayOptions;
+import com.google.android.gms.maps.model.TileProvider;
+
+import java.io.File;
+
+import android.os.Environment;
+import android.database.sqlite.SQLiteDatabase;
+import android.database.Cursor;
+
+/**
+ * Created by Christoph Lambio on 30/03/2018.
+ * Based on AirMapLocalTileManager.java
+ * Copyright (c) zavadpe
+ */
+
+public class AirMapMbTile extends AirMapFeature {
+
+    class AIRMapMbTileProvider implements TileProvider {
+        private static final int BUFFER_SIZE = 16 * 1024;
+        private int tileSize;
+        private String pathTemplate;
+
+
+        public AIRMapMbTileProvider(int tileSizet, String pathTemplate) {
+            this.tileSize = tileSizet;
+            this.pathTemplate = pathTemplate;
+        }
+
+        @Override
+        public Tile getTile(int x, int y, int zoom) {
+            byte[] image = readTileImage(x, y, zoom);
+            return image == null ? TileProvider.NO_TILE : new Tile(this.tileSize, this.tileSize, image);
+        }
+
+        public void setPathTemplate(String pathTemplate) {
+            this.pathTemplate = pathTemplate;
+        }
+
+        public void setTileSize(int tileSize) {
+            this.tileSize = tileSize;
+        }
+
+        private byte[] readTileImage(int x, int y, int zoom) {
+            String rawQuery = "SELECT * FROM map INNER JOIN images ON map.tile_id = images.tile_id WHERE map.zoom_level = {z} AND map.tile_column = {x} AND map.tile_row = {y}";
+
+            try {
+                SQLiteDatabase offlineDataDatabase = SQLiteDatabase.openDatabase(this.pathTemplate, null, SQLiteDatabase.OPEN_READONLY);
+                String query = rawQuery.replace("{x}", Integer.toString(x))
+                        .replace("{y}", Integer.toString(y))
+                        .replace("{z}", Integer.toString(zoom));
+                Cursor cursor = offlineDataDatabase.rawQuery(query, null);
+                if(cursor.moveToFirst()){
+                    byte[] tile = cursor.getBlob(5);
+                    cursor.close();
+                    offlineDataDatabase.close();
+                    return tile;
+                }
+                offlineDataDatabase.close();
+                return null;
+            } catch (Exception e) {
```

Comment:
I believe catching a generic exception like this might lead to some bugs. In general, it is better to be specific about your exceptions when you know they exist, like [this](https://developer.android.com/reference/android/database/sqlite/SQLiteException.html).

Consider the following:
1. You open a connection successfully, but your string replace (currently line line 55) throws an error. I believe you will end up not closing the database.

I would look into other code that uses this lib and see how they do error handling, but on very cursory glance, I would structure it like the following:

```

try {
  SQLiteDatabase offlineDataDatabase offlineDataDatabase = SQLiteDatabase.openDatabase(this.pathTemplate, null, SQLiteDatabase.OPEN_READONLY);
  String query = rawQuery.replace("{x}", Integer.toString(x)).replace("{y}", Integer.toString(y))
                                                                                                    .replace("{z}", Integer.toString(zoom));
  Cursor cursor = offlineDataDatabase.rawQuery(query, null);
  byte[] tile = null
  if(cursor.moveToFirst()) {
    byte[] tile = cursor.getBlob(5);
    cursor.close();
  } 
  offlineDataDatabase.close();
  return tile;
} catch (SQLiteCantOpenDatabaseException e) {
  // error handle, this case is when database can't open so return null?
} catch (android.database.sqlite.SQLiteException e) {
  // error handle, past an open database connection so close and return null.
} catch (Exception e) {
  // all other cases, return null?
}


```

Reply:
Possibly even better:

```
byte[] tile = null
try {
  SQLiteDatabase offlineDataDatabase offlineDataDatabase = SQLiteDatabase.openDatabase(this.pathTemplate, null, SQLiteDatabase.OPEN_READONLY);
  String query = rawQuery.replace("{x}", Integer.toString(x)).replace("{y}", Integer.toString(y))
                                                                                                    .replace("{z}", Integer.toString(zoom));
  Cursor cursor = offlineDataDatabase.rawQuery(query, null);
  if(cursor.moveToFirst()) {
    tile = cursor.getBlob(5);
    cursor.close();
  } 
  offlineDataDatabase.close();
} catch (SQLiteCantOpenDatabaseException e) {
  // error handle, this case is when database can't open so return null?
} catch (android.database.sqlite.SQLiteException e) {
  // error handle, past an open database connection so close and return null.
} catch (Exception e) {
  // all other cases, return null?
}

return tile;
```

Reply:
Thanks for pointing that out. I changed the "closing the database" problem and included some `catch`. I prefer the code written slightly different, so I left that. I included a more elaborated error handling namely `SQLiteCantOpenDatabaseException`, `SQLiteDatabaseCorruptException` and `SQLiteDatabaseLockedException`.